File: javascript/grade-school/grade-school.js
export class GradeSchool {

    #grades;

    constructor() {
        this.#grades = {};
    }

    roster() {
        return structuredClone(this.#grades);
    }

    add(student, gradeNumber) {
        if (!this.#grades[gradeNumber]) {
            this.#grades[gradeNumber] = [];
        }

        Object.keys(this.#grades)
            .forEach((grade) => 
                this.#grades[grade] = this.#grades[grade]
                                .filter(stdnt => stdnt !== student)
            );

        this.#grades[gradeNumber].push(student);
        this.#grades[gradeNumber].sort();
    }

    grade(gradeNumber) {
        return this.#grades[gradeNumber] 
            ? [...this.#grades[gradeNumber]] : [];
    }
}

File: python/grade-school/grade_school.py
class School:
    def __init__(self):
        self._roster = {}
        self._added = []

    def add_student(self, name, grade):

        if name not in self.roster():
            self._roster[grade] = self._roster.get(grade, []) + [name]
            self._added.append(True)
        else:
            self._added.append(False)

    def roster(self):
        return [student
                for grade in sorted(self._roster.keys())
                for student in sorted(self._roster[grade])]

    def grade(self, grade_number):
        return sorted(self._roster.get(grade_number, []))

    def added(self):
        return self._added
